File: frontend/src/types/resume.ts
// RenderCV v2.8 Schema Types

export type SocialNetworkName =
  | "LinkedIn"
  | "GitHub"
  | "GitLab"
  | "IMDB"
  | "Instagram"
  | "ORCID"
  | "Mastodon"
  | "StackOverflow"
  | "ResearchGate"
  | "YouTube"
  | "Google Scholar"
  | "Telegram"
  | "WhatsApp"
  | "Leetcode"
  | "X"
  | "Bluesky"
  | "Reddit";

export interface SocialNetwork {
  network: SocialNetworkName;
  username: string;
}

// --- Entry Types ---

export interface EducationEntry {
  institution: string;
  area: string;
  degree?: string;
  start_date?: string;
  end_date?: string;
  date?: string;
  location?: string;
  summary?: string;
  highlights?: string[];
}

export interface ExperienceEntry {
  company: string;
  position: string;
  start_date?: string;
  end_date?: string;
  date?: string;
  location?: string;
  summary?: string;
  highlights?: string[];
}

export interface NormalEntry {
  name: string;
  start_date?: string;
  end_date?: string;
  date?: string;
  location?: string;
  summary?: string;
  highlights?: string[];
}

export interface PublicationEntry {
  title: string;
  authors: string[];
  doi?: string;
  url?: string;
  journal?: string;
  date?: string;
  summary?: string;
}

export interface OneLineEntry {
  label: string;
  details: string;
}

export interface BulletEntry {
  bullet: string;
}

export type SectionEntry =
  | EducationEntry
  | ExperienceEntry
  | NormalEntry
  | PublicationEntry
  | OneLineEntry
  | BulletEntry
  | string;

export type SectionType =
  | "education"
  | "experience"
  | "projects"
  | "skills"
  | "publications"
  | "certificates"
  | "awards"
  | "extracurricular"
  | "custom";

// --- CV Data ---

export interface CvData {
  name?: string;
  headline?: string;
  location?: string;
  email?: string;
  phone?: string;
  website?: string;
  social_networks?: SocialNetwork[];
  sections?: Record<string, SectionEntry[]>;
}

// --- Design Config ---

export interface PageConfig {
  size?: string;
  top_margin?: string;
  bottom_margin?: string;
  left_margin?: string;
  right_margin?: string;
}

export interface ColorsConfig {
  body?: string;
  name?: string;
  headline?: string;
  connections?: string;
  section_titles?: string;
  links?: string;
  footer?: string;
}

export interface FontSizeConfig {
  body?: string;
  name?: string;
  headline?: string;
  connections?: string;
  section_titles?: string;
}

export interface TypographyConfig {
  font_family?: string;
  line_spacing?: string;
  alignment?: string;
  font_size?: FontSizeConfig;
}

export interface LinksConfig {
  underline?: boolean;
  show_external_link_icon?: boolean;
}

export interface HeaderConfig {
  alignment?: string;
}

export interface SectionTitlesConfig {
  type?: string;
}

export interface DesignConfig {
  theme?: string;
  page?: PageConfig;
  colors?: ColorsConfig;
  typography?: TypographyConfig;
  links?: LinksConfig;
  header?: HeaderConfig;
  section_titles?: SectionTitlesConfig;
}

// --- Top-Level Resume Data ---

export interface ResumeData {
  cv?: CvData;
  design?: DesignConfig;
}

// --- Default Resume ---

export const DEFAULT_RESUME: ResumeData = {
  cv: {
    name: "John Doe",
    headline: "Software Engineer",
    location: "San Francisco, CA",
    email: "john.doe@example.com",
    phone: "+1 (555) 123-4567",
    website: "https://johndoe.dev",
    social_networks: [
      { network: "LinkedIn", username: "johndoe" },
      { network: "GitHub", username: "johndoe" },
    ],
    sections: {
      education: [
        {
          institution: "University of California, Berkeley",
          area: "Computer Science",
          degree: "B.S.",
          start_date: "2016-09",
          end_date: "2020-05",
          location: "Berkeley, CA",
          highlights: [
            "Dean's List, Fall 2018 and Spring 2019",
            "GPA: 3.8/4.0",
          ],
        } as EducationEntry,
      ],
      experience: [
        {
          company: "Acme Corp",
          position: "Software Engineer",
          start_date: "2020-06",
          end_date: "present",
          location: "San Francisco, CA",
          summary: "Full-stack development for enterprise SaaS platform.",
          highlights: [
            "Built a real-time notification system serving 50k+ users",
            "Reduced API response times by 40% through query optimization",
          ],
        } as ExperienceEntry,
      ],
      skills: [
        { label: "Languages", details: "Python, TypeScript, Go, SQL" } as OneLineEntry,
        { label: "Frameworks", details: "React, Next.js, FastAPI, Django" } as OneLineEntry,
        { label: "Tools", details: "Docker, Kubernetes, AWS, Git" } as OneLineEntry,
      ],
    },
  },
  design: {
    theme: "classic",
  },
};

File: backend/app/models/resume.py
from __future__ import annotations

from enum import Enum
from typing import Any, Optional, Union

from pydantic import BaseModel, ConfigDict


# --- Social Network Enum ---

class SocialNetworkName(str, Enum):
    LINKEDIN = "LinkedIn"
    GITHUB = "GitHub"
    GITLAB = "GitLab"
    IMDB = "IMDB"
    INSTAGRAM = "Instagram"
    ORCID = "ORCID"
    MASTODON = "Mastodon"
    STACKOVERFLOW = "StackOverflow"
    RESEARCHGATE = "ResearchGate"
    YOUTUBE = "YouTube"
    GOOGLE_SCHOLAR = "Google Scholar"
    TELEGRAM = "Telegram"
    WHATSAPP = "WhatsApp"
    LEETCODE = "Leetcode"
    X = "X"
    BLUESKY = "Bluesky"
    REDDIT = "Reddit"


class SocialNetwork(BaseModel):
    network: SocialNetworkName
    username: str


# --- Entry Types ---

class EducationEntry(BaseModel):
    institution: str
    area: str
    degree: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    date: Optional[str] = None
    location: Optional[str] = None
    summary: Optional[str] = None
    highlights: Optional[list[str]] = None


class ExperienceEntry(BaseModel):
    company: str
    position: str
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    date: Optional[str] = None
    location: Optional[str] = None
    summary: Optional[str] = None
    highlights: Optional[list[str]] = None


class NormalEntry(BaseModel):
    name: str
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    date: Optional[str] = None
    location: Optional[str] = None
    summary: Optional[str] = None
    highlights: Optional[list[str]] = None


class PublicationEntry(BaseModel):
    title: str
    authors: list[str]
    doi: Optional[str] = None
    url: Optional[str] = None
    journal: Optional[str] = None
    date: Optional[str] = None
    summary: Optional[str] = None


class OneLineEntry(BaseModel):
    label: str
    details: str


class BulletEntry(BaseModel):
    bullet: str


# TextEntry is represented as a plain string in the sections list,
# so we don't need a separate model for it.

# Union of all entry types. Plain strings represent TextEntry.
EntryType = Union[
    EducationEntry,
    ExperienceEntry,
    NormalEntry,
    PublicationEntry,
    OneLineEntry,
    BulletEntry,
    str,
]


# --- Design Sub-Models ---

class PageConfig(BaseModel):
    model_config = ConfigDict(extra="allow")

    size: Optional[str] = None
    top_margin: Optional[str] = None
    bottom_margin: Optional[str] = None
    left_margin: Optional[str] = None
    right_margin: Optional[str] = None


class ColorsConfig(BaseModel):
    model_config = ConfigDict(extra="allow")

    body: Optional[str] = None
    name: Optional[str] = None
    headline: Optional[str] = None
    connections: Optional[str] = None
    section_titles: Optional[str] = None
    links: Optional[str] = None
    footer: Optional[str] = None


class FontSizeConfig(BaseModel):
    model_config = ConfigDict(extra="allow")

    small: Optional[str] = None
    normal: Optional[str] = None
    large: Optional[str] = None


class TypographyConfig(BaseModel):
    model_config = ConfigDict(extra="allow")

    font_family: Optional[str] = None
    font_size: Optional[FontSizeConfig] = None
    line_spacing: Optional[str] = None
    alignment: Optional[str] = None


class LinksConfig(BaseModel):
    model_config = ConfigDict(extra="allow")

    underline: Optional[bool] = None
    show_external_link_icon: Optional[bool] = None


class HeaderConfig(BaseModel):
    model_config = ConfigDict(extra="allow")

    alignment: Optional[str] = None


class SectionTitlesConfig(BaseModel):
    model_config = ConfigDict(extra="allow")

    type: Optional[str] = None


class EntriesConfig(BaseModel):
    model_config = ConfigDict(extra="allow")

    date_and_location_width: Optional[str] = None


class DesignConfig(BaseModel):
    model_config = ConfigDict(extra="allow")

    theme: Optional[str] = "classic"
    page: Optional[PageConfig] = None
    colors: Optional[ColorsConfig] = None
    typography: Optional[TypographyConfig] = None
    links: Optional[LinksConfig] = None
    header: Optional[HeaderConfig] = None
    section_titles: Optional[SectionTitlesConfig] = None
    entries: Optional[EntriesConfig] = None


# --- CV Data ---

class CvData(BaseModel):
    name: Optional[str] = None
    headline: Optional[str] = None
    location: Optional[str] = None
    email: Optional[Union[str, list[str]]] = None
    phone: Optional[Union[str, list[str]]] = None
    website: Optional[Union[str, list[str]]] = None
    social_networks: Optional[list[SocialNetwork]] = None
    sections: Optional[dict[str, list[Any]]] = None


# --- Top-Level Resume Data ---

class LocaleConfig(BaseModel):
    model_config = ConfigDict(extra="allow")

    language: str = "english"


class ResumeData(BaseModel):
    cv: CvData
    design: Optional[DesignConfig] = None
    locale: Optional[LocaleConfig] = None
    settings: Optional[dict[str, Any]] = None
